specs/core/1_shard-data-chains.md

```diff
@@ -501,395 +252,484 @@ Define a `SubkeyReveal` as follows:
 
-For each `response` in `block.body.branch_responses`, if `response.responding_to_custody_challenge == False`, run:
+For each `reveal` in `block.body.early_subkey_reveals`, run the following function:
 
 ```python
-def process_branch_exploration_response(state: BeaconState,
-                                        response: BranchResponse) -> None:
-    challenge = get_branch_challenge_record_by_id(response.challenge_id)
-    assert verify_merkle_branch(
-        leaf=response.data,
-        branch=response.branch,
-        depth=challenge.depth,
-        index=challenge.data_index,
-        root=challenge.root
-    )
-    # Must wait at least ENTRY_EXIT_DELAY before responding to a branch challenge
-    assert get_current_epoch(state) >= challenge.inclusion_epoch + ENTRY_EXIT_DELAY
-    state.branch_challenge_records.pop(challenge)
-    # Reward the proposer
-    proposer_index = get_beacon_proposer_index(state, state.slot)
-    state.validator_balances[proposer_index] += base_reward(state, index) // MINOR_REWARD_QUOTIENT
-```
-
-If `response.responding_to_custody_challenge == True`, run:
+def process_subkey_reveal(state: BeaconState,
+                          reveal: SubkeyReveal) -> None:
+    assert verify_custody_subkey_reveal(reveal)
 
-```python
-def process_branch_custody_response(state: BeaconState,
-                                    response: BranchResponse) -> None:
-    challenge = get_custody_challenge_record_by_id(response.challenge_id)
-    responder = state.validator_registry[challenge.responder_index]
-    # Verify we're not too late
-    assert get_current_epoch(state) < responder.withdrawable_epoch
-    # Verify the Merkle branch *of the data tree*
-    assert verify_merkle_branch(
-        leaf=response.data,
-        branch=response.branch,
-        depth=challenge.max_depth,
-        index=challenge.offset,
-        root=challenge.data_root
+    # Either the reveal is of a future period, or it is of the current period and the validator is still active
+    is_early_reveal = reveal.period > get_current_custody_period(state) or (
+        reveal.period == get_current_custody_period(state) and
+        state.validator_registry[reveal.validator_index].exit_epoch > get_current_epoch(state)
     )
-    # Responder wins
-    if hash(challenge.responder_subkey + response.data) == challenge.current_custody_tree_node:
-        penalize_validator(state, challenge.challenger_index, challenge.responder_index)
-    # Challenger wins
+
+    if is_early_reveal:
+        assert state.validator_registry[reveal.validator_index].slashed_epoch > get_current_epoch(state)
+        slash_validator(state, reveal.validator_index, reveal.revealer_index)
+        state.validator_balances[reveal.revealer_index] += base_reward(state, index) // MINOR_REWARD_QUOTIENT
+    elif reveal.period == state.validator_registry[reveal.validator_index].next_subkey_to_reveal and reveal.mask == ZERO_HASH:
```

Comment:
By the variable names, it seems "Subkey reveals" section only processes `early_subkey_reveals`; are they all in `is_early_reveal` cases? Or it's better to adjust the variable naming?